Responsive side nav

robdodson opened the pull request:
Fixes #1371

Note: If this is written content for the site please include the issue number from your [content proposal](https://github.com/GoogleChrome/web.dev/issues?q=is%3Aissue+is%3Aopen+label%3A%22content+proposal%22).

Changes proposed in this pull request:

- Adds a Header component.
- Adds a SideNav component.
- Adds wicg-inert polyfill.
- Adds :focus-visible polyfill.
- Defines actions for opening and closing the SideNav, as well as inerting other parts of the page.

samthor commented on src/lib/components/Header/_styles.scss:
```diff
@@ -0,0 +1,132 @@
+@import '../../../styles/tools/breakpoints';
+@import '../../../styles/tools/mixins';
+
+web-header:not(:defined) {
```
does this work in a polyfilled universe?

robdodson replied:
probably not. I also think that any FOUC with this header is bad so I'm going to move all of its guts to the light dom

samthor commented on src/lib/components/Header/index.js:
```diff
@@ -0,0 +1,67 @@
+
+  firstUpdated() {
+    this.hamburgerBtn = this.querySelector(".web-header__hamburger-btn");
+    this.hamburgerBtn.addEventListener("click", openSideNav);
```
Shouldn't this be a `@click=...` part of the rendered HTML?

robdodson replied:
keeping this as is because I moved all the children back to light dom.

samthor replied:
sure. I wonder if it's worth calling out `BaseElement` subclasses which don't have a `render()` method.

i.e., there's no value to `BaseElement` in this case?

robdodson replied:
yeah maybe I should just switch to extending LitElement.

samthor commented on src/lib/components/SideNav/index.js:
```diff
@@ -0,0 +1,199 @@
+  }
+
+  addEventListeners() {
+    this.addEventListener("click", this.onCloseSideNav);
```
For everything not on `this`, same as my other comment—shouldn't this be done inside lit?

We could even add a top-level element that's the same size that handles the other on... handlers.

robdodson replied:
I fixed this for the handler referencing sideNavContainerEl. But I kept the addListeners block for the stuff pointing at `this`. I explored creating a wrapper element, but [lit's syntax for passing options to a declarative handler](https://github.com/Polymer/lit-html/pull/531) seems to make the code more confusing than it's worth.

Basically I was ending up creating objects that contained the touch code, and the functions within those objects have to be bound back to `this` so it's just kinda gross.

samthor commented on src/lib/components/SideNav/index.js:
```diff
@@ -0,0 +1,199 @@
+
+    const oldVal = this.expanded_;
+    this.animatable = true;
+    this.addEventListener("transitionend", this.onTransitionEnd);
```
`{once: true}`?

samthor replied:
And then it can just be inline handler?

robdodson replied:
Added the `once: true` object and removed the `removeEventListener` in the handler. But I didn't inline this listener. See above for my reasons re: lit's options object.

robdodson commented:
@samthor did some updates based on your feedback. Would you mind taking a look at the two commits?

Along with the stuff I commented on I also:
- Made the header fixed position to match what devsite does.
- Added margin styles to `<main>` to account for fixed position header.
- Tweaked z-index for a few elements. I'm planning to create a separate PR to introduce an actual z-index scale so we don't end up with a bunch of arbitrary values. But didn't want to complicate this PR.
- Added an `overflow: hidden` class to body during the `openSideNav()`/`closeSideNav()` actions.

samthor commented on rollup.config.js:
```diff
@@ -18,6 +18,7 @@ module.exports = [
       format: "iife",
       entryFileNames: outputPattern,
       chunkFileNames: outputPattern,
+      sourcemap: true,
```
this seems unrelated but sure 👍

samthor commented on src/lib/actions.js:
```diff
@@ -1,6 +1,7 @@
 import {store} from "./store";
 import {saveUserUrl} from "./fb";
 import {runLighthouse, fetchReports} from "./lighthouse-service";
+import "wicg-inert";
```
I'm not against inert but an alternative approach is that focusing outside the nav bar implicitly closes it (it's what we do on Santa). Just my 2c and doesn't require the polyfill.

robdodson replied:
I don't think that would prevent a screen reader from breaking outside of the side nav tho. Are you also marking the other areas as aria-hidden?

samthor commented on src/lib/components/SideNav/index.js:
```diff
@@ -0,0 +1,195 @@
+    this.animatable = true;
+    this.addEventListener("transitionend", this.onTransitionEnd, {once: true});
+    this.expanded_ = val;
+    if (this.expanded_) {
```
why not `this.inert = !this.expanded_;` ?

samthor commented on src/lib/components/SideNav/index.js:
```diff
@@ -0,0 +1,195 @@
+
+  disconnectedCallback() {
+    super.disconnectedCallback();
+    store.unsubscribe(this.onStateChanged);
```
The way I read the docs for `firstUpdated`, it seems to be one-time, not after every addition to the page.

https://lit-element.polymer-project.org/guide/lifecycle#firstupdated

Maybe move the subscribe call to `connectedCallback`?

samthor commented on src/lib/components/Header/index.js:
```diff
@@ -0,0 +1,34 @@
+    this.hamburgerBtn.addEventListener("click", expandSideNav);
+  }
+
+  manageFocus() {
```
Maybe comment where this is called from.

Alternatively, is there any precedent to overloading `.focus()` itself?

robdodson replied:
ermmmmmm I feel a little sketchy about overloading focus... I'll add a comment instead :P
